File: editor/rbtree/rbtree.c
/*

  rbtree.c

  Implements the Red-Black Tree algorithm as described in
  "Introduction to Algorithms"

  Copyright 2009-2020 Matthew T. Pandina. All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions
  are met:

  1. Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.

  2. Redistributions in binary form must reproduce the above copyright
  notice, this list of conditions and the following disclaimer in the
  documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY MATTHEW T. PANDINA "AS IS" AND ANY
  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL MATTHEW T. PANDINA OR
  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  SUCH DAMAGE.

*/

#include <stdlib.h>
#include <string.h>

#include "rbtree.r"

void rbtree_init(rbtree_t *self,
                 rbtree_node_t *nil,
                 unsigned int rbtree_node_t_size,
                 int (*CompareFunc)(const rbtree_node_t *,
                                    const rbtree_node_t *)) {
  self->nil = nil;
  self->nil->parent = self->nil->left = self->nil->right = self->nil;
  self->nil->color = RBTREE_NODE_COLOR_BLACK;

  self->root = self->nil;
  self->rbtree_node_t_size = rbtree_node_t_size;
  self->Compare = CompareFunc;
}

void rbtree_destroy(rbtree_t *self) {
  memset(self, 0, sizeof(rbtree_t));
}

rbtree_node_t *rbtree_search(rbtree_t *self, rbtree_node_t *k) {
  rbtree_node_t *x = self->root;
  while (x != self->nil && self->Compare(k, x)) {
    if (self->Compare(k, x) < 0)
      x = x->left;
    else
      x = x->right;
  }
  return x;
}

void rbtree_insert(rbtree_t *self, rbtree_node_t *z) {
  rbtree_node_t *y = self->nil;
  rbtree_node_t *x = self->root;
  while (x != self->nil) {
    y = x;
    if (self->Compare(z, x) < 0)
      x = x->left;
    else
      x = x->right;
  }
  z->parent = y;
  if (y == self->nil)
    self->root = z;
  else {
    if (self->Compare(z, y) < 0)
      y->left = z;
    else
      y->right = z;
  }
  z->left = self->nil;
  z->right = self->nil;
  z->color = RBTREE_NODE_COLOR_RED;
  InsertFixup(self, z);
}

rbtree_node_t *rbtree_minimum(rbtree_t *self) {
  return Minimum(self, self->root);
}

rbtree_node_t *rbtree_maximum(rbtree_t *self) {
  return Maximum(self, self->root);
}

rbtree_node_t *rbtree_predecessor(rbtree_t *self, rbtree_node_t *x) {
  return Predecessor(self, x);
}

rbtree_node_t *rbtree_successor(rbtree_t *self, rbtree_node_t *x) {
  return Successor(self, x);
}

rbtree_node_t *rbtree_delete(rbtree_t *self, rbtree_node_t *z) {
  rbtree_node_t *x, *y;
  if (z->left == self->nil || z->right == self->nil)
    y = z;
  else
    y = Successor(self, z);
  if (y->left != self->nil)
    x = y->left;
  else
    x = y->right;
  x->parent = y->parent;
  if (y->parent == self->nil)
    self->root = x;
  else {
    if (y == y->parent->left)
      y->parent->left = x;
    else
      y->parent->right = x;
  }
  if (y != z && self->rbtree_node_t_size > sizeof(rbtree_node_t))
    memcpy(((unsigned char *)z) + sizeof(rbtree_node_t),
           ((unsigned char *)y) + sizeof(rbtree_node_t),
           self->rbtree_node_t_size - sizeof(rbtree_node_t));
  if (y->color == RBTREE_NODE_COLOR_BLACK)
    DeleteFixup(self, x);
  return y;
}

static void InorderTreeWalk(rbtree_t *self,
                            rbtree_node_t *x,
                            void (*ApplyFunc)(rbtree_node_t *, void *),
                            void *context) {
  if (x != self->nil) {
    InorderTreeWalk(self, x->left, ApplyFunc, context);
    ApplyFunc(x, context);
    InorderTreeWalk(self, x->right, ApplyFunc, context);
  }
}

void rbtree_inorderwalk(rbtree_t *self,
                        void (*ApplyFunc)(rbtree_node_t *, void *),
                        void *context) {
  InorderTreeWalk(self, self->root, ApplyFunc, context);
}

static void PreorderTreeWalk(rbtree_t *self,
                             rbtree_node_t *x,
                             void (*ApplyFunc)(rbtree_node_t *, void *),
                             void *context) {
  if (x != self->nil) {
    ApplyFunc(x, context);
    PreorderTreeWalk(self, x->left, ApplyFunc, context);
    PreorderTreeWalk(self, x->right, ApplyFunc, context);
  }
}

void rbtree_preorderwalk(rbtree_t *self,
                         void (*ApplyFunc)(rbtree_node_t *, void *),
                         void *context) {
  PreorderTreeWalk(self, self->root, ApplyFunc, context);
}

static void PostorderTreeWalk(rbtree_t *self,
                              rbtree_node_t *x,
                              void (*ApplyFunc)(rbtree_node_t *, void *),
                              void *context) {
  if (x != self->nil) {
    PostorderTreeWalk(self, x->left, ApplyFunc, context);
    PostorderTreeWalk(self, x->right, ApplyFunc, context);
    ApplyFunc(x, context);
  }
}

void rbtree_postorderwalk(rbtree_t *self,
                          void (*ApplyFunc)(rbtree_node_t *, void *),
                          void *context) {
  PostorderTreeWalk(self, self->root, ApplyFunc, context);
}

File: editor/rbtree/rbtree.r
/*

  rbtree.r

  Implements the Red-Black Tree algorithm as described in
  "Introduction to Algorithms"

  Note: This is a private header file, which should only be included
  by the header files of extensions to the Red-Black Tree
  implementation which need to call these inline functions. For
  example: rbtree+setinsert.h should include this header, but main.c
  should not.

  Copyright 2009-2020 Matthew T. Pandina. All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions
  are met:

  1. Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.

  2. Redistributions in binary form must reproduce the above copyright
  notice, this list of conditions and the following disclaimer in the
  documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY MATTHEW T. PANDINA "AS IS" AND ANY
  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL MATTHEW T. PANDINA OR
  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  SUCH DAMAGE.

*/

#pragma once

#include "rbtree.h"

static inline void LeftRotate(rbtree_t *self, rbtree_node_t *x) {
  rbtree_node_t *y = x->right;
  x->right = y->left;
  if (y->left != self->nil)
    y->left->parent = x;
  y->parent = x->parent;
  if (x->parent == self->nil)
    self->root = y;
  else {
    if (x == x->parent->left)
      x->parent->left = y;
    else
      x->parent->right = y;
  }
  y->left = x;
  x->parent = y;
}

static inline void RightRotate(rbtree_t *self, rbtree_node_t *y) {
  rbtree_node_t *x = y->left;
  y->left = x->right;
  if (x->right != self->nil)
    x->right->parent = y;
  x->parent = y->parent;
  if (y->parent == self->nil)
    self->root = x;
  else {
    if (y == y->parent->right)
      y->parent->right = x;
    else
      y->parent->left = x;
  }
  x->right = y;
  y->parent = x;
}

static inline void InsertFixup(rbtree_t *self, rbtree_node_t *z) {
  rbtree_node_t *y;
  while (z->parent->color == RBTREE_NODE_COLOR_RED) {
    if (z->parent == z->parent->parent->left) {
      y = z->parent->parent->right;
      if (y->color == RBTREE_NODE_COLOR_RED) {
        z->parent->color = RBTREE_NODE_COLOR_BLACK;
        y->color = RBTREE_NODE_COLOR_BLACK;
        z->parent->parent->color = RBTREE_NODE_COLOR_RED;
        z = z->parent->parent;
      } else {
        if (z == z->parent->right) {
          z = z->parent;
          LeftRotate(self, z);
        }
        z->parent->color = RBTREE_NODE_COLOR_BLACK;
        z->parent->parent->color = RBTREE_NODE_COLOR_RED;
        RightRotate(self, z->parent->parent);
      }
    } else {
      y = z->parent->parent->left;
      if (y->color == RBTREE_NODE_COLOR_RED) {
        z->parent->color = RBTREE_NODE_COLOR_BLACK;
        y->color = RBTREE_NODE_COLOR_BLACK;
        z->parent->parent->color = RBTREE_NODE_COLOR_RED;
        z = z->parent->parent;
      } else {
        if (z == z->parent->left) {
          z = z->parent;
          RightRotate(self, z);
        }
        z->parent->color = RBTREE_NODE_COLOR_BLACK;
        z->parent->parent->color = RBTREE_NODE_COLOR_RED;
        LeftRotate(self, z->parent->parent);
      }
    }
  }
  self->root->color = RBTREE_NODE_COLOR_BLACK;
}

static inline rbtree_node_t *Minimum(rbtree_t *self, rbtree_node_t *x) {
  while (x->left != self->nil)
    x = x->left;
  return x;
}

static inline rbtree_node_t *Maximum(rbtree_t *self, rbtree_node_t *x) {
  while (x->right != self->nil)
    x = x->right;
  return x;
}

static inline rbtree_node_t *Predecessor(rbtree_t *self, rbtree_node_t *x) {
  if (x->left != self->nil)
    return Maximum(self, x->left);
  rbtree_node_t *y = x->parent;
  while (y != self->nil && x == y->left) {
    x = y;
    y = y->parent;
  }
  return y;
}

static inline rbtree_node_t *Successor(rbtree_t *self, rbtree_node_t *x) {
  if (x->right != self->nil)
    return Minimum(self, x->right);
  rbtree_node_t *y = x->parent;
  while (y != self->nil && x == y->right) {
    x = y;
    y = y->parent;
  }
  return y;
}

static inline void DeleteFixup(rbtree_t *self, rbtree_node_t *x) {
  rbtree_node_t *w;
  while (x != self->root && x->color == RBTREE_NODE_COLOR_BLACK) {
    if (x == x->parent->left) {
      w = x->parent->right;
      if (w->color == RBTREE_NODE_COLOR_RED) {
        w->color = RBTREE_NODE_COLOR_BLACK;
        x->parent->color = RBTREE_NODE_COLOR_RED;
        LeftRotate(self, x->parent);
        w = x->parent->right;
      }
      if (w->left->color == RBTREE_NODE_COLOR_BLACK &&
	  w->right->color == RBTREE_NODE_COLOR_BLACK) {
        w->color = RBTREE_NODE_COLOR_RED;
        x = x->parent;
      } else {
        if (w->right->color == RBTREE_NODE_COLOR_BLACK) {
          w->left->color = RBTREE_NODE_COLOR_BLACK;
          w->color = RBTREE_NODE_COLOR_RED;
          RightRotate(self, w);
          w = x->parent->right;
        }
        w->color = x->parent->color;
        x->parent->color = RBTREE_NODE_COLOR_BLACK;
        w->right->color = RBTREE_NODE_COLOR_BLACK;
        LeftRotate(self, x->parent);
        x = self->root;
      }
    } else {
      w = x->parent->left;
      if (w->color == RBTREE_NODE_COLOR_RED) {
        w->color = RBTREE_NODE_COLOR_BLACK;
        x->parent->color = RBTREE_NODE_COLOR_RED;
        RightRotate(self, x->parent);
        w = x->parent->left;
      }
      if (w->right->color == RBTREE_NODE_COLOR_BLACK &&
	  w->left->color == RBTREE_NODE_COLOR_BLACK) {
        w->color = RBTREE_NODE_COLOR_RED;
        x = x->parent;
      } else {
        if (w->left->color == RBTREE_NODE_COLOR_BLACK) {
          w->right->color = RBTREE_NODE_COLOR_BLACK;
          w->color = RBTREE_NODE_COLOR_RED;
          LeftRotate(self, w);
          w = x->parent->left;
        }
        w->color = x->parent->color;
        x->parent->color = RBTREE_NODE_COLOR_BLACK;
        w->left->color = RBTREE_NODE_COLOR_BLACK;
        RightRotate(self, x->parent);
        x = self->root;
      }
    }
  }
  x->color = RBTREE_NODE_COLOR_BLACK;
}
